src/Searchbar.vue

```diff
@@ -15,38 +15,95 @@ export default {
+      }
+
+      function someMatch(info, regexes) {
+        return regexes.some((regex) => regex.test(info));
+      }
+
       if (!this.data) {
         return undefined;
       }
-      const matches = [];
-      const regex = new RegExp(this.value, 'i');
+      let matches = [];
+      const regexes = this.value.split(' ').map(keyword => new RegExp(keyword, 'i'));
       this.data.forEach((entry) => {
         const { headings, src, title } = entry;
         const keywords = entry.keywords || '';
         let hasMatchingHeading = false;
+        const headingEntries = [];
+
         Object.entries(headings).forEach(([id, text]) => {
-          if (regex.test(text)) {
-            hasMatchingHeading = true;
-            matches.push({
-              heading: { id, text },
-              keywords,
-              src,
-              title,
-            });
-          }
-        });
-        if (!hasMatchingHeading) {
-          if (regex.test(title) || regex.test(keywords)) {
-            matches.push(entry);
+            const info = [title].concat(keywords).concat(text).join(' ');
+            if (allMatch(info, regexes)) {
+              if(someMatch(text, regexes)) {
+                hasMatchingHeading = true;
+              }
+              headingEntries.push({
+                heading: { id, text },
+                keywords,
+                src,
+                title,
+              });
+            }
+          });
+          if (!hasMatchingHeading) {
+            const info = [title].concat(keywords).join(' ');
+            if(allMatch(info, regexes)) {
+              matches.push(entry);
+            }
           }
-        }
+          matches = matches.concat(headingEntries);
       });
       return matches;
     },
   },
   filters: {
     highlight(value, phrase) {
-      return value.replace(new RegExp(`(${phrase})`, 'gi'), '<mark>$1</mark>');
+      function mergeIntervals(intervals) {
+        const stack = [];
+        let top = null;
+        if (intervals.length <= 1) {
+          return intervals;
+        }
+        const sortedIntervals = intervals.sort((a, b) => a.start - b.start);
+
+        stack.push(sortedIntervals[0]);
+        for (let i = 1; i < sortedIntervals.length; i += 1) {
+          top = stack[stack.length - 1];
+
+          if (top.end < sortedIntervals[i].start) {
+            stack.push(sortedIntervals[i]);
+          } else if (top.end < sortedIntervals[i].end) {
+            top.end = sortedIntervals[i].end;
+          }
+        }
+
+        return stack;
+      }
+
+      const keywords = phrase.split(' ').filter(keyword => keyword !== '');
+      let matchPos = [];
+      let match;
+      keywords.forEach((keyword) => {
+        const regex = new RegExp(`(${keyword})`, 'gi');
+        match = regex.exec(value);
+        while (match != null) {
+          matchPos.push({ start: match.index, end: match.index + keyword.length });
+          match = regex.exec(value);
+        }
+      });
```

Comment:
line 86-96 can be a method of its own (trying to apply SLAP here).

Then line 98 can simply be `matchPos = mergeIntervals(findAllIntervals(value, searchPhrase))`